src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -143,38 +145,64 @@ void reconnect() throws TimeoutException, ExecutionException, InterruptedExcepti
         // Synchronize here instead of method signature to make mockito work without deadlocking
         synchronized (this) {
             logger.atInfo().log("Reconnecting MQTT client most likely due to device configuration change");
-            disconnect();
+            disconnect().get(getTimeout(), TimeUnit.MILLISECONDS);
             connect().get(getTimeout(), TimeUnit.MILLISECONDS);
         }
     }
 
-    @SuppressWarnings("PMD.NullAssignment")
     protected synchronized CompletableFuture<Boolean> connect() {
-        if (connection != null) {
+        // future not done indicates an ongoing connect attempt, caller should wait on that future
+        // instead of starting another connect attempt.
+        if (connectionFuture != null && !connectionFuture.isDone()) {
+            return connectionFuture;
+        }
+        // We're already connected, there's nothing to do
+        if (currentlyConnected.get()) {
             return CompletableFuture.completedFuture(true);
         }
+        // For the initial connect, client connects with cleanSession=true and disconnects.
+        // This deletes any previous session information maintained by IoT Core.
+        // For subsequent connects, the client connects with cleanSession=false
+        CompletableFuture<Void> voidCompletableFuture =  CompletableFuture.completedFuture(null);
+        if (initalConnect.get()) {
+            voidCompletableFuture = establishConnection(true).thenCompose((session) -> {
+                initalConnect.set(false);
+                return disconnect();
+            });
+        }
+
+        connectionFuture = voidCompletableFuture.thenCompose((b) -> establishConnection(false))
+                .thenApply((sessionPresent) -> {
+                    currentlyConnected.set(true);
+                    logger.atInfo().kv("sessionPresent", sessionPresent)
+                            .log("Successfully connected to AWS IoT Core");
+                    if (!sessionPresent) {
+                        resubscribe();
+                    }
```

Comment:
we now need to resubscribe no matter what because the session which might be present won't have the subscriptions that we want.

Reply:
I don't think the logic changed to get to the state "if session present won't have the subscriptions that we want".

Reply:
Perhaps you're right because of the initialConnection bool